src/shogun/base/SGObject.h

```diff
@@ -1011,5 +1029,85 @@ class CSGObject
 	/** Subscriber used to call onNext, onComplete etc.*/
 	SGSubscriber* m_subscriber_params;
 };
+
+#ifndef SWIG
+#ifndef DOXYGEN_SHOULD_SKIP_THIS
+namespace sgo_details
+{
+template <typename T1, typename T2>
+bool dispatch_array_type(const CSGObject* obj, const std::string& name, T2&& lambda)
+{
+	Tag<CDynamicObjectArray*> tag_array_sg(name);
+	if (obj->has(tag_array_sg))
+	{
+		auto dispatched = obj->get(tag_array_sg);
+		lambda(*dispatched); // is stored as a pointer
+		return true;
+	}
+
+	Tag<std::vector<T1*>> tag_vector(name);
+	if (obj->has(tag_vector))
+	{
+		auto dispatched = obj->get(tag_vector);
+		lambda(dispatched);
+		return true;
+	}
+
+	return false;
+}
+
+struct GetByName
+{
+};
+
+struct GetByNameIndex
+{
+	GetByNameIndex(index_t index) : m_index(index) {}
+	index_t m_index;
+};
+
+template <typename T>
+CSGObject* get_if_possible(const CSGObject* obj, const std::string& name, GetByName)
+{
+	return obj->has<T*>(name) ? obj->get<T*>(name) : nullptr;
+}
+
+template <typename T>
+CSGObject* get_if_possible(const CSGObject* obj, const std::string& name, GetByNameIndex how)
+{
+	CSGObject* result = nullptr;
+	try
+	{
+		// there is no "has" that checks for array types, so check implicitly
+		result = obj->get<T>(name, how.m_index);
+	}
+	catch (const std::exception&) {}
+	return result;
+}
+
+template<typename T>
+CSGObject* get_dispatch_all_base_types(const CSGObject* obj, const std::string& name,
+		T&& how)
+{
+	if (auto* result = get_if_possible<CKernel>(obj, name, how))
```

Comment:
@gf712 I would love to centralise this list via using some structure of base types we have in shogun. Was there something in place for that already?

Reply:
I was trying to get them all in this [file](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/lib/sg_types.h).

Reply:
ok thanks will add that later, but the PR is too big already anyways :)

Reply:
@karlnapf Why did you not add CDistance and CLabels?

Reply:
Forgot! Thanks. Will send a fix soon.

Reply:
actually it would be good to have a dispatcher for base types, just like we have for `Any`.
@geektoni added `CMachine` in another PR recently, and I am sure there will be dispatching happening at other places in the future as well.
I mean for now this is the centralised place which is re-used at least.
Maybe we can write an entrance task, and label it for the detox project?